docs/containerization-design.rst

```diff
@@ -0,0 +1,358 @@
+
+#. Provide plumbing for a fine-grained Canary system that can allow Azkaban/jobtypes and platform full
+   control of ramping up their binaries, independent of each other.
+
+Future Extensions
+-----------------
+1. Provide the fine-grained Canary system for Multiple components that make up Azkaban to help in their
+   independent evolution.
+
+High Level Design Summary
+*************************
+
+.. image:: docs/figures/containerized-high-level-arch.png
+
+1. Azkaban will follow a **Disposable Container** model. This implies that whenever a flow is to be launched, the
+   **dispatch logic** will launch a fresh Pod and the pod is destroyed at the conclusion of the flow.
+#. Isolation is achieved per flow (Not at job level). Jobs/subflows that are a part of a flow, will run within the
+   confines of the pod launched for orchestrating the flow. Job level isolation was explored and rejected:
+
+   * It is very disruptive given the existing Azkaban architecture. Major portions of executor code will have to be
+     re-written to accommodate job level isolation.
+   * It appears too resource hungry to launch a separate pod per job. Separate container within the same pod is
+     possible. But again, this would have required rewriting major parts of flow - job wiring. This is something
+     that can be reconsidered in future.
+
+#. The pod will be launched with default compute/memory resources, but override parameters will be available to
+   request more resources for the flow orchestration.
+#. For this design iteration, the web server will stay outside of k8s. This does not preclude the need for the
+   web server to talk to flow pods to fetch logs or send control commands (Such as Cancel). To enable this
+   communication, an Envoy Proxy based Ingress Controller is introduced, which will allow the web server to
+   communicate with Flow Pods. There is no need to set node ports for flow pods.
+#. In order to satisfy `key Requirement #3 <#Key-Requirements-for-Containerization>`_, the execution environment
+   for flow pods will be constructed dynamically at run-time.
+
+   * Azkaban will provide a mechanism to dynamically select versions of components that constitute a functional
+     Azkaban Executor environment at dispatch time.
+   * Following this, a series of init containers will pull intended versions of various components to compose the
+     complete execution environment.
+   * The dynamic selection process will be utilized to provide canary capability for various Azkaban components.
+   * The design also introduces a few Admin APIs to make the task of image management easier.
+
+Design Details
+**************
+
+Image Management
+----------------
+* Azkaban will rely on docker images to create execution environment for flows. In order to satisfy
+  `key Requirement #3 <#Key-Requirements-for-Containerization>`_, the final container image will be constructed
```

Comment:
Consider replacing `container image` with `pod` as that's more accurate. I.e  `the final pod will be constructed dynamically...`

Reply:
Done